src/include/migraphx/op/scatter.hpp

```diff
@@ -33,29 +44,38 @@ struct scatter
         return {{"normalize_axes", normalize}};
     }
 
-    std::string name() const { return "scatter"; }
-
     shape normalize_compute_shape(std::vector<shape> inputs) const
     {
         check_shapes{inputs, *this}.has(3).standard();
-        return inputs.front();
+        // If non-packed, this converts to a packed output while preserving permutation of tensor
+        return inputs.front().with_lens(inputs.front().lens());
     }
 
     argument compute(const shape& output_shape, std::vector<argument> args) const
     {
         argument result{output_shape};
-        // max dimension in axis
+        auto& self = static_cast<const Derived&>(*this);
+
+        // max dimension in each axis
         auto axis_dim_size = output_shape.lens()[axis];
+        // cast all arguments as correct type
         visit_all(result, args[0], args[2])([&](auto output, auto data, auto update) {
+            // copy all of data to output
             std::copy(data.begin(), data.end(), output.begin());
             args[1].visit([&](auto indices) {
                 auto ind_s = indices.get_shape();
+                // iterate through items in index
                 shape_for_each(ind_s, [&](const auto& idx) {
-                    auto out_idx  = idx;
-                    auto index    = indices[ind_s.index(idx)];
+                    auto out_idx = idx;
+                    auto index   = indices[ind_s.index(idx)];
```

Comment:
Doing `indices[ind_s.index(idx)]` is incorrect(which was done previously). The problem is that `.index()` returns an offset, but the `[]` accepts the position(not the offset). This should do `indices(idx.begin(), idx.end())` instead.

Actually, it would be even better to add overloads to tensor_view to take a range:

```cpp
    template<class Range, MIGRAPHX_REQUIRES(not std::is_integral<Range>{})>
    T& operator[](Range&& r)
    {
        return (*this)(r.begin(), r.end());
    }

    template<class Range, MIGRAPHX_REQUIRES(not std::is_integral<Range>{})>
    const T& operator[](Range&& r) const
    {
        return (*this)(r.begin(), r.end());
    }
```

Then you could just write `indices[idx]` instead.

Reply:
> Doing `indices[ind_s.index(idx)]` is incorrect(which was done previously). The problem is that `.index()` returns an offset, but the `[]` accepts the position(not the offset). This should do `indices(idx.begin(), idx.end())` instead.
> 

You'll have to clarify.  It passes the tests the way it is.



> Actually, it would be even better to add overloads to tensor_view to take a range:
> 
> ```c++
>     template<class Range, MIGRAPHX_REQUIRES(not std::is_integral<Range>{})>
>     T& operator[](Range&& r)
>     {
>         return (*this)(r.begin(), r.end());
>     }
> 
>     template<class Range, MIGRAPHX_REQUIRES(not std::is_integral<Range>{})>
>     const T& operator[](Range&& r) const
>     {
>         return (*this)(r.begin(), r.end());
>     }
> ```
> 
> Then you could just write `indices[idx]` instead.

I think that's out of scope for this issue.

Reply:
> You'll have to clarify. It passes the tests the way it is.

I think the tests are only using standard shape(which the offset and position are the same), but if you use non-standard shape this will break.

Ideally, in the future, I would like to have a some kind of tagged type to catch these errors at compile-time because I think there are other places where we are doing this. Of course, a tagged type checking is not for this PR though.

> I think that's out of scope for this issue.

Sure, I'll open a PR for these overloads.

Reply:
I'm not sure we're all on the same page.  Testing for nonstandard shapes conflicts with Shucai's comment [here, above](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/pull/1124#discussion_r824999498) that we should assume standard shape (and the current code does.)  Am I misreading this, or is this introducing a change to a specified behavior that we've been planning all along? 

It's not even clear to me that using a scatter operation on a nonstandard tensor shape even makes sense from a user perspective, or if it's only helpful when starting with contiguous memory.  

Maybe this would be a good discussion topic for this afternoon's team meeting, if we can keep it short.

Reply:
After more discussion, I'll try adding the suggested [] overloads to the tensor_view definition and see if it still passes existing tests.  In the interest of limiting the scope and time of this PR, if it needs further work to pass then we'll back out the change and write a new issue.

Also, will postpone creating tests using nonstandard shape inputs, and write a separate issue to be addressed as part of the GPU implementation of scatter/gather operators.

Reply:
Didn't add the [] overloads; created [issue 1132](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/issues/1132#issue-1168999696) instead

Reply:
> I'll try adding the suggested []

When you add the [] overloads then you can change `indices[ind_s.index(idx)]` to `indices[idx]` so nonstandard index calculation should be correct.